### ISSUE TITLE
[Bug] CKAN.TransactionalKraken: Registry already enlisted

### ISSUE THREAD

**Issue by ohthebiomesyoullgoooooooo** [KSP-CKAN/CKAN#3539]
Background
----------
* Operating System: Windows 10
* CKAN Version: Hubble
* KSP Version: 1.12.0
* Have you made any manual changes to your GameData folder (i.e., not via CKAN)? nope

Problem
----------
Everytime i try to install a visual mod it says the code down there, is it the program? or just my bad pc..

**Describe the bug**

**Steps to reproduce**
* Go to CKAN
* Click on any visual mod
* See error CKAN.TransactionalKraken

**Expected behavior**
if i try to install any visual mod like realplume stock, EVE, Astronomer, etc.
it says the code down there

Log file: [CKAN.log](https://github.com/KSP-CKAN/CKAN/files/8189403/CKAN.log)

```
CKAN.TransactionalKraken: Registry already enlisted with tx 2e0d1b0a-82e1-4941-9452-1cfdaaa8c252:10, can't enlist with tx 2e0d1b0a-82e1-4941-9452-1cfdaaa8c252:14
   at CKAN.Registry.EnlistWithTransaction()
   at CKAN.Registry.RegisterModule(CkanModule mod, IEnumerable`1 absolute_files, GameInstance ksp, Boolean autoInstalled)
   at CKAN.ModuleInstaller.Install(CkanModule module, Boolean autoInstalled, Registry registry, HashSet`1& possibleConfigOnlyDirs, String filename)
   at CKAN.ModuleInstaller.InstallList(ICollection`1 modules, RelationshipResolverOptions options, RegistryManager registry_manager, HashSet`1& possibleConfigOnlyDirs, IDownloader downloader, Boolean ConfirmPrompt)
   at CKAN.Main.InstallMods(Object sender, DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.OnDoWork(DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.WorkerThreadStart(Object argument)
```

**Comment by HebaruSan** [comment]
See #3502 and #3517 for previous reports that we couldn't continue investigating after the authors stopped responding.

@ohthebiomesyoullgoooooooo, this comment has a link to a test download and instructions for how to use it:

https://github.com/KSP-CKAN/CKAN/issues/3517#issuecomment-1014788274

Please try that and share a link to the log file. Thanks!

**Comment by ohthebiomesyoullgoooooooo** [comment]
Roger.

**Comment by ohthebiomesyoullgoooooooo** [comment]
kinda

**Comment by ohthebiomesyoullgoooooooo** [comment]
I got the error again, see below.
![2022-03-03](https://user-images.githubusercontent.com/91695802/156604835-d5195bd6-af46-49a7-b94a-43ed641aa8be.png)

**Comment by HebaruSan** [comment]
Yes, that's expected. The point was to generate a log file so we could investigate this.

**Comment by ohthebiomesyoullgoooooooo** [comment]
also BOTH??? where?????????????

**Comment by ohthebiomesyoullgoooooooo** [comment]
wait i just had a realisation..

**Comment by ohthebiomesyoullgoooooooo** [comment]
WAIT A SECOND HOL UP THE MODS THAT I TRIED TO INSTALL ARE THERE??? what? there was a error, AND THEY ARE THERE IN MY  GAMEDATA FOLDER??

**Comment by HebaruSan** [comment]
Please share a link to the log file.

**Comment by ohthebiomesyoullgoooooooo** [comment]
how am i supposed to do it?

**Comment by HebaruSan** [comment]
This comment has a link to a test download and instructions for how to use it:

https://github.com/KSP-CKAN/CKAN/issues/3517#issuecomment-1014788274

**Comment by ohthebiomesyoullgoooooooo** [comment]
Also, for some reason... Mods work now? I can install outer planets. but ill try

**Comment by ohthebiomesyoullgoooooooo** [comment]
and google detected it

**Comment by HebaruSan** [comment]
@ohthebiomesyoullgoooooooo, these comments are not helping us to investigate this issue. We need a log file as requested three times. If you're able to provide this, please do, and we can re-open this.

**Comment by ohthebiomesyoullgoooooooo** [comment]
Okay, i can provide it.  i got it working and i need to create the bug.

**Comment by ohthebiomesyoullgoooooooo** [comment]
I'm trying to get the bug created..

**Comment by ohthebiomesyoullgoooooooo** [comment]
I GOT THE BUG, IM SENDNG THE LOG FILE

**Comment by ohthebiomesyoullgoooooooo** [comment]
how do i send it?

**Comment by ohthebiomesyoullgoooooooo** [comment]
<?xml version="1.0" encoding="utf-8" ?>
<log4net>
    <appender name="CkanFile" type="log4net.Appender.RollingFileAppender">
        <file value="./CKAN.log" />
        <appendToFile value="true" />
        <rollingStyle value="Size" />
        <maximumFileSize value="50MB" />
        <maxSizeRollBackups value="30" />
        <lockingModel type="log4net.Appender.FileAppender+MinimalLock" />

        <filter type="log4net.Filter.StringMatchFilter">
            <stringToMatch value="Finding latest available" />
            <acceptOnMatch value="false" />
        </filter>
        <filter type="log4net.Filter.StringMatchFilter">
            <stringToMatch value="Installable: " />
            <acceptOnMatch value="false" />
        </filter>
        <filter type="log4net.Filter.StringMatchFilter">
            <regexToMatch value="Checking .*\.\.\." />
            <acceptOnMatch value="false" />
        </filter>
        <filter type="log4net.Filter.StringMatchFilter">
            <stringToMatch value="Got contents: " />
            <acceptOnMatch value="false" />
        </filter>
        <filter type="log4net.Filter.StringMatchFilter">
            <stringToMatch value="Got owners: " />
            <acceptOnMatch value="false" />
        </filter>

        <layout type="log4net.Layout.PatternLayout">
            <conversionPattern value="%date{yyyy-MM-dd HH:mm:ss} %level %logger - %message%newline" />
        </layout>
    </appender>

    <root>
        
        <level value="WARN" />
        <appender-ref ref="CkanFile" />
    </root>

    
    <logger name="CKAN.Registry">
        <level value="DEBUG" />
    </logger>
    <logger name="CKAN.ModuleInstaller">
        <level value="DEBUG" />
    </logger>
    <logger name="CKAN.RegistryManager">
        <level value="DEBUG" />
    </logger>
    <logger name="CKAN.GameInstance">
        <level value="DEBUG" />
    </logger>
    <logger name="CKAN.Repo">
        <level value="DEBUG" />
    </logger>
    <logger name="CKAN.Main">
        <level value="DEBUG" />
    </logger>
</log4net>

**Comment by HebaruSan** [comment]
Good try, but that's the wrong file. You want the **`CKAN.log`** file.

**Comment by ohthebiomesyoullgoooooooo** [comment]
Sorry.

**Comment by ohthebiomesyoullgoooooooo** [comment]
I found a log in CKAN.txt
2022-03-04 18:41:53 DEBUG CKAN.Registry - Already enlisted with tx 8de9ee4f-6ee1-4c21-916a-5339f1f08402:1

**Comment by ohthebiomesyoullgoooooooo** [comment]
Is this good?

**Comment by HebaruSan** [comment]
That's closer! But we need the whole file. You can attach it to a comment by clicking this in the bottom of the comment box:

![image](https://user-images.githubusercontent.com/1559108/156858598-c011dbb0-9428-460f-9bd4-85e6aef6cb59.png)

**Comment by ohthebiomesyoullgoooooooo** [comment]
Be warned, i got a lot of downloads.

**Comment by ohthebiomesyoullgoooooooo** [comment]
[CKAN.log](https://github.com/KSP-CKAN/CKAN/files/8189403/CKAN.log)

**Comment by ohthebiomesyoullgoooooooo** [comment]
wait do i have to do with the debugging one or just my regular one?

**Comment by ohthebiomesyoullgoooooooo** [comment]
that is the debugging one

**Comment by ohthebiomesyoullgoooooooo** [comment]
Hello? 
are you there? or are you just looking in it?

**Comment by HebaruSan** [comment]
I'm eating dinner and watching _The Expanse_ with my wife, which will take at least another 45 minutes.
But thanks, this is the right file.

**Comment by HebaruSan** [comment]
It looks like the transaction is calling `Registry.Prepare` but not `Registry.Commit`.

```
2022-03-04 18:53:03 INFO CKAN.ModuleInstaller - Installed RecoveryController 0.0.4.1
2022-03-04 18:53:03 DEBUG CKAN.Registry - Registering module RecoveryController 0.0.4.1
2022-03-04 18:53:03 DEBUG CKAN.Registry - Already enlisted with tx fb315af8-b63d-489e-b72a-cd266f542342:8
2022-03-04 18:53:03 INFO CKAN.RegistryManager - Saving CKAN registry at C:/Games/Kerbal Space Program/CKAN\registry.json
2022-03-04 18:53:21 DEBUG CKAN.Registry - Registry prepared to commit transaction
2022-03-04 18:56:12 DEBUG CKAN.Registry - Started with MakingHistory-DLC 1.12.0, BreakingGround-DLC 1.7.0, xScienceContinued 6.0.0.9, TextureReplacer v4.5.3, DistantObject v2.1.1.6, PlanetShine 0.2.6.6, Scatterer-sunflare 3:v0.0831, PoodsCalmNebulaSkybox v1.3.0, ClickThroughBlocker 1:0.1.10.17, ZeroMiniAVC 1:1.1.0.2, ToolbarController 1:0.1.9.6, SpaceTuxLibrary 0.0.8.4, DistantObject-default v2.1.1.6, PlanetShine-Config-Default 0.2.6.6, CameraTools v1.20.0, MechJeb2 2.12.3.0, Toolbar 1:1.8.0.7, HotSpot 0.9.0, WarpDrive 0.9.4.2, ReentryParticleEffect 1.9.1.1, DeadlyReentry 1:v.7.9.0, HullcamVDSContinued 0.2.2, CactEyeCommunityRefocused 1.5.2.7, RecoveryController 0.0.4.1, CustomBarnKit 1.1.22.0, KSP-Recall v0.2.2.1, Trajectories v2.4.3, FMRSContinued 1.2.9.2, TarsierSpaceTechnologyWithGalaxies 1:7.13, KAS 1.10, KIS 1.29, CommunityCategoryKit v5.2.0.0, RealisticAtmospheres 1.3.2, Kopernicus 2:release-1.12.1-83, ModularFlightIntegrator 1.2.10.0, OuterPlanetsMod 2:2.2.10, CommunityTerrainTexturePack 1:1.0.4, CommNetAntennasExtension 2.1.6, ScienceSituationInfo 1:1.3.5, CommunityPartsTitles 0.9.1, CommNetAntennasInfo 3.1.2, CommNetAntennasConsumptor 3.5.5, RemoteTechRedevAntennas 0.1.1, TweakScale v2.4.6.8, SpaceXLegs 2.9.2, removing ClickThroughBlocker, and keeping MakingHistory-DLC 1.12.0, BreakingGround-DLC 1.7.0, xScienceContinued 6.0.0.9, TextureReplacer v4.5.3, DistantObject v2.1.1.6, PlanetShine 0.2.6.6, Scatterer-sunflare 3:v0.0831, PoodsCalmNebulaSkybox v1.3.0, ZeroMiniAVC 1:1.1.0.2, ToolbarController 1:0.1.9.6, SpaceTuxLibrary 0.0.8.4, DistantObject-default v2.1.1.6, PlanetShine-Config-Default 0.2.6.6, CameraTools v1.20.0, MechJeb2 2.12.3.0, Toolbar 1:1.8.0.7, HotSpot 0.9.0, WarpDrive 0.9.4.2, ReentryParticleEffect 1.9.1.1, DeadlyReentry 1:v.7.9.0, HullcamVDSContinued 0.2.2, CactEyeCommunityRefocused 1.5.2.7, RecoveryController 0.0.4.1, CustomBarnKit 1.1.22.0, KSP-Recall v0.2.2.1, Trajectories v2.4.3, FMRSContinued 1.2.9.2, TarsierSpaceTechnologyWithGalaxies 1:7.13, KAS 1.10, KIS 1.29, CommunityCategoryKit v5.2.0.0, RealisticAtmospheres 1.3.2, Kopernicus 2:release-1.12.1-83, ModularFlightIntegrator 1.2.10.0, OuterPlanetsMod 2:2.2.10, CommunityTerrainTexturePack 1:1.0.4, CommNetAntennasExtension 2.1.6, ScienceSituationInfo 1:1.3.5, CommunityPartsTitles 0.9.1, CommNetAntennasInfo 3.1.2, CommNetAntennasConsumptor 3.5.5, RemoteTechRedevAntennas 0.1.1, TweakScale v2.4.6.8, SpaceXLegs 2.9.2; our dlls are 666-ModuleManagerWatchDog, Interstellar-Redist, ModuleManager, KerbalEngineer
```

The line after "prepared to commit" should start with "Committing registry tx" (and that pattern is present in several other places in the log), but instead we have a 2.9 minute delay followed by some `FindReverseDependencies` instrumentation.

Checking the documentation for explanations, maybe other enlistees called `PreparingEnlistment.ForceRollback`, or maybe an exception was thrown?

- https://docs.microsoft.com/en-us/dotnet/api/system.transactions.transactionscope.complete?view=netframework-4.7.2
- https://docs.microsoft.com/en-us/dotnet/api/system.transactions.ienlistmentnotification.prepare?view=netframework-4.7.2#system-transactions-ienlistmentnotification-prepare(system-transactions-preparingenlistment)
- https://docs.microsoft.com/en-us/dotnet/api/system.transactions.preparingenlistment.prepared?view=netframework-4.7.2

**Comment by ohthebiomesyoullgoooooooo** [comment]
oh, so how long will it take until it gets fixed on me?

**Comment by ohthebiomesyoullgoooooooo** [comment]
i'll have to install those files.

**Comment by ohthebiomesyoullgoooooooo** [comment]
maybe.

**Comment by HebaruSan** [comment]
Conversation locked due to repeated unproductive comments and resultant notification spam. We are analyzing the log file, and we don't know how long that will take. If that allows us to determine a code fix, we will create a pull request for it. If we need more information, we will create a new debug component to generate more logs and unlock the conversation.

**Comment by HebaruSan** [comment]
For background, the transactions in the log all start with the same lower case UUID, then end with a colon and an incrementing number. The one noted above that apparently never ends is `:8`, which is an attempt to install some mods.

The previous transaction, `:7`, removes some mods. Interestingly, it has a lot of errors about missing files:

```
2022-03-04 18:49:45 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/EnvironmentalVisualEnhancements : Could not find file 'C:\Games\Kerbal Space Program\GameData\EnvironmentalVisualEnhancements'.
2022-03-04 18:49:45 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/EnvironmentalVisualEnhancements/button.png : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\EnvironmentalVisualEnhancements\button.png'.
2022-03-04 18:49:45 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/EnvironmentalVisualEnhancements/EnvironmentalVisualEnhancements.version : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\EnvironmentalVisualEnhancements\EnvironmentalVisualEnhancements.version'.
[...]
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/AstronomersVisualPack : Could not find file 'C:\Games\Kerbal Space Program\GameData\AstronomersVisualPack'.
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/AstronomersVisualPack/AstronomersVisualPack.version : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\AstronomersVisualPack\AstronomersVisualPack.version'.
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/AstronomersVisualPack/AVP_Configs : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\AstronomersVisualPack\AVP_Configs'.
[...]
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/Scatterer : Could not find file 'C:\Games\Kerbal Space Program\GameData\Scatterer'.
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/Scatterer/icon : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\Scatterer\icon'.
2022-03-04 18:50:04 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/Scatterer/icon/icon.png : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\Scatterer\icon\icon.png'.
[...]
2022-03-04 18:50:05 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/999_KSP-Recall : Could not find file 'C:\Games\Kerbal Space Program\GameData\999_KSP-Recall'.
2022-03-04 18:50:05 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/999_KSP-Recall/CHANGE_LOG.md : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\999_KSP-Recall\CHANGE_LOG.md'.
2022-03-04 18:50:05 INFO CKAN.ModuleInstaller - Failure in locating file C:/Games/Kerbal Space Program/GameData/999_KSP-Recall/KSP-Recall.cfg : Could not find a part of the path 'C:\Games\Kerbal Space Program\GameData\999_KSP-Recall\KSP-Recall.cfg'.
```

So maybe an unstated step in reproducing this is to install mods with CKAN, then delete some of those files manually before uninstalling them in CKAN? Too bad the steps to reproduce weren't described (and presumably the "manual changes" question in the issue template wasn't answered accurately).

**Comment by HebaruSan** [comment]
The `Failure in locating...` message comes from here, presumably thrown by `File.GetAttributes`:

https://github.com/KSP-CKAN/CKAN/blob/46e989046980183772f5ae03dad65ec8ff43c80f/Core/ModuleInstaller.cs#L724-L771

Might be good to probe that a bit more. If the `file_transaction.Delete` line is throwing instead, then maybe some internal state of the transaction system could be getting corrupted for the next transaction?

**Comment by HebaruSan** [comment]
This is probably but not certainly fixed in the v1.31.0 release. Closing for now; if it still happens with the latest CKAN, please let us know and we can re-open this.
